CONTRIBUTING.txt

```diff
@@ -0,0 +1,98 @@
+    3. To preserve and improve the code quality while encouraging appropriate experimentation and risk-taking
+    4. To minimize dependence on individual contributors by encouraging a large pool of active contributors
+    5. To be inclusive of many viewpoints and to harness a diverse set of skills
+    6. To provide an encouraging environment where contributors learn and improve their skills
+    7. To protect the free and open nature of the FreeCAD project
+
+2. Fundamentals
+
+    1. FreeCAD uses the git distributed revision control system.
+    2. Source code for the main application and related subprojects is hosted on github.com in the FreeCAD organization, hereafter referred to as 'the platform'.
+    3. Problems are discrete, well-defined limitations or bugs.
+    4. FreeCAD uses the platform's issue-tracking system to track problems
+    5. Patches are sets of code changes that resolve a single problem
+    6. FreeCAD uses the Pull Request workflow for evaluating and accepting patches
+
+
+3. Roles
+    1. "User": An member of the wider FreeCAD community who uses the software.
+    2. "Contributor":  A person who submits a patch that resolves a previously identified problem. Contributors do not have commit access to the repository unless they are also maintainers. Everyone, without distinction or discrimination, SHALL have an equal right to become a Contributor
+    3. "Maintainer": A person who merges patches. Maintainers may or may not be Contributors. Their role is to enforce the process. Maintainers have commit access to the repository.
+    4. "Administrator": Administrators have additional authority to maintain the list of designated maintainers.
+
+4. Licensing, Ownership, and Credit
+    1. FreeCAD uses the Lesser General Public License, version 2, or superior (LGPL2+) for all source code.  Additional details can be found in the LICENSE file.
+    2. All contributions to FreeCAD SHALL use the same license as the project.
+    3. All contributions are owned by their authors unless assigned to another.
+    4. FreeCAD does not have a mandatory copyright assignment policy.
+    5. A Contributor who wishes to be identified in the Credits section of the application "About" dialog is responsible for identifying themselves appropriately by notifying an administrator or maintainer.
+
+5. Contribution Requirements
+
+    1. Patches are submitted in the form of Pull Requests (PR)
+    2. Maintainers and Contributors MUST have a platform account and SHOULD use their real names or a well-known alias. If the platform alias differs from the account alias on the FreeCAD Forum, effort SHOULD be taken to avoid confusion.
+    3. A PR SHOULD be a minimal and accurate answer to exactly one identified and agreed-on problem.
+    4. A PR SHOULD refrain from adding additional dependencies to the FreeCAD project unless no other option is available.
+    5. Code submissions MUST adhere to the code style guidelines of the project if these are defined.
```

Comment:
This was somewhat contentious in the developer meeting.  Everyone agrees that a style guide is a good thing. Not everyone agrees that compliance should be required.
This walks the line and says compliance is required IF a guide is defined.  
First, should we publish a guide?  Where?
Should the "MUST" be stated as a "SHOULD"?

Reply:
Personally I believe that this should stay a "MUST" to make life much easier for all of the people who have to read/review/revise the code in the future.  Otherwise we are building in a lot of "friction" for everyone who follows the original writer.

There is a different issue about when/if/how to enforce new requirements for a style in already existing code.  I am in favor of "the next time someone touches this file, the first thing they must do is format it to comply with the style guide".  It causes some up-front work to save work afterwards.

Reply:
I also think that a single coding style for each language for all of FreeCAD should be a "MUST",  I would expect that to be quite controversial :-)

Reply:
> First, should we publish a guide? Where?

clang-format has been in the codebase some months now. So in the case of C++ the should and where are done. Just the details need agreement.

Reply:
> clang-format has been in the codebase some months now.

I was referring more to a human readable policy or guide.  I've educated a lot of people about PEP8 and Black formatting over the years. Having a document I could point to would be helpful.   Is it the same with clang-format?  Do developers need to be clued in how to use it?  
   
Ideally there would be a top-level guide that applies project-wide and if workbenches extend or depart from that, they define it in the Mod folder.

Reply:
Is this a separate problem to resolve?  Should we open an issue on setting and documenting code style guides?

Reply:
clang-format is a long list of specs. Each is generally quite readable. To create a full descriptive document AND maintain it in sync will be a fair amount of work, but would be useful in terms of explaining the reasoning behind the definitions. But equally, depending on your tools you might be able to hover over the entries and get a verbose description (works in CLion).

Reply:
I'm speaking about policy, not specification.  When it comes to python, the spec is basically encoded in PEP8 so our policy can simply say something like "all python should be formatted according to PEP8"
However, that may not be enough.  We may need additional style rules to cover project-specific stuff like error messages, translation strings, etc.  
What I'm asking here, is whether we should split the discussion of formatting policy to a new issue.  Then this part of the conversation can be resolved and the document reads the same:  follow the style guideline if it exists.

Reply:
I've created a new issue to continue the discussion around code formatting.
https://github.com/FreeCAD/FreeCAD/issues/7650

If anyone has more to contribute to the policy here, please offer a text suggestion for discussion or or I'll resolve this conversation and move on.